ulib/libax/src/cbindings/socket.rs

```diff
@@ -0,0 +1,337 @@
+        let res_socket_addr = socket_addr as *mut ctypes::sockaddr_in;
+
+        let a = socket.lock().recvfrom(buf);
+        a.map(|res| {
+            if let Some(addr) = res.1 {
+                (*res_socket_addr).sin_family = ctypes::AF_INET as u16;
+                (*res_socket_addr).sin_addr = in_addr {
+                    s_addr: u32::from_be_bytes(addr.addr.as_bytes().try_into().unwrap()).to_be(),
+                };
+                (*res_socket_addr).sin_port = addr.port.to_be();
+                *socket_len = size_of::<sockaddr>() as u32;
+            }
+            res.0
+        })
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_recv(
+    socket_fd: c_int,
+    buf_ptr: *mut c_void,
+    len: ctypes::size_t,
+    _flag: c_int, //no effect
+) -> ctypes::ssize_t {
+    ax_call_body!(ax_recv, {
+        if buf_ptr.is_null() {
+            return Err(LinuxError::EINVAL);
+        }
+        let socket = Filelike::from_fd(socket_fd)?.into_socket()?;
+        let buf = core::slice::from_raw_parts_mut(buf_ptr as *mut u8, len);
+        let a = socket.lock().recv(buf)?;
+        Ok(a)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_listen(
+    socket_fd: c_int,
+    _backlog: c_int, // no effect
+) -> ctypes::ssize_t {
+    ax_call_body!(ax_listen, {
+        let socket = Filelike::from_fd(socket_fd)?.into_socket()?;
+        socket.lock().listen()?;
+        Ok(0)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_accept(
+    socket_fd: c_int,
+    socket_addr: *mut ctypes::sockaddr,
+    socket_len: *mut ctypes::socklen_t,
+) -> ctypes::ssize_t {
+    ax_call_body!(ax_accept, {
+        if socket_addr.is_null() || socket_len.is_null() {
+            return Err(LinuxError::EINVAL);
+        }
+        let socket = Filelike::from_fd(socket_fd)?.into_socket()?;
+        let ressocket = socket.lock().accept()?;
+        let addr = ressocket.peer_addr()?;
+        let fd = add_new_socket(Socket::Tcp(ressocket)).ok_or(LinuxError::ENFILE)?;
+        let mut res_socket_addr = *(socket_addr as *mut ctypes::sockaddr_in);
+        res_socket_addr.sin_family = ctypes::AF_INET as u16;
+        res_socket_addr.sin_addr = in_addr {
+            s_addr: u32::from_be_bytes(addr.addr.as_bytes().try_into().unwrap()).to_be(),
+        };
+        res_socket_addr.sin_port = addr.port.to_be();
+        *socket_len = size_of::<sockaddr>() as u32;
+        Ok(fd)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_shutdown(
+    socket_fd: c_int,
+    _flag: c_int, // no effect
+) -> ctypes::ssize_t {
+    ax_call_body!(ax_shutdown, {
+        let socket = Filelike::from_fd(socket_fd)?.into_socket()?;
+        socket.lock().shutdown()?;
+        Ok(0)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_dnsquery(
+    node: *const c_char,
+    addr: *mut ctypes::sockaddr,
+    len: ctypes::size_t,
+) -> c_int {
+    let addr_slice = slice_from_raw_parts_mut(addr as *mut sockaddr_in, len);
```

Comment:
Why use `slice_from_raw_parts_mut` instead of `core::slice::from_raw_parts_mut`?